tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+	---@param src_pdf string
+	---@param page_dimensions table
+	---@param tolerance number
+	---@param force boolean
+	---@return integer[] matching_pages
+	---@return integer[] failed_pages
+	---@return string pdf_version
+	---@overload fun(src_pdf:string, page_dimensions:table, tolerance:number, force:boolean): nil, string?, nil
+	check_dimensions = function(src_pdf, page_dimensions, tolerance, force)
+		local pdf
+		if kpse.in_name_ok_silent_extended(src_pdf) then
+			pdf = pdfe.open(src_pdf)
+		else
+			return nil, ("Opening %s not permitted."):format(src_pdf)
+		end
+
+		-- collect which pages succeded the dimension check
+		local succ = {}
+		-- collect which pages failed the dimension check
+		local failed = {}
+		for _, i in ipairs(page_dimensions) do
+			local p = i.page
+			local page = pdfe.getpage(pdf, p)
+			if not page then
+				-- page not found -> skip it
+				table.insert(failed, {page=i, reason="not found"})
+			else
+				local mediabox = pdfe.getbox(page, "MediaBox")
+				local w = bp2pt(mediabox[3] - mediabox[1])
+				local h = bp2pt(mediabox[4] - mediabox[2])
+				if math.abs(w - i.width) > tolerance or math.abs(h - i.height) > tolerance and not force then
+					table.insert(failed, {page=i, reason="dimension", real_width=w, real_height=h})
+				else
+					table.insert(succ, p)
+				end
+			end
+		end
+
+		local v_major, v_minor = pdfe.getversion(pdf)
+		local pdf_version = ("%d.%d"):format(v_major, v_minor)
+
+		pdfe.close(pdf)
+
+		table.sort(succ)
+		table.sort(failed, function(a,b) return a.page.page < b.page.page end)
+		return succ, failed, pdf_version
+	end
+end
+
+-- restrict the complete rest of the script by undefining security relevant libraries
+-- this defines an allow-list what functions of these libraries still should be accessible
+local env = {
+	-- lua libraries
+	arg      = arg,
+	ipairs   = ipairs,
+	math     = math,
+	os       = { type = os.type, },
+	pairs    = pairs,
+	print    = print,
+	table    = table,
+	tonumber = tonumber,
+	tostring = tostring,
+	select   = select,
+
+	-- luatex specific libraries
+	lfs      = {isfile=lfs.isfile},
+	kpse     = kpse,
+
+	-- memoize-extract specific global
+	STAGE    = STAGE,
+}
+
+local exit
+if STAGE == "testing" then
+	-- in testing environment avoid exiting the whole test
+	-- -> instead raise an error which can be catched
+
+	-- store the error function independent of the environment
+	local error = error
+	exit = {
+		error = function() error("exited with error") end,
+		warn  = function() error("exited with warn") end,
+		succ  = function() error("exited with succ") end,
+	}
+else
+	local os_exit = os.exit
+	exit = {
+		error = function() os_exit(11) end,
+		warn  = function() os_exit(10) end,
+		succ  = function() os_exit(0) end,
+	}
+end
+
+do
+	-- I don't like using the debug library, but getting a traceback here is a
+	-- must to find where the error originates from
+	local debug_traceback = debug.traceback
```

Comment:
`debug.traceback` is totally fine, and makes debugging much easier, so good idea!